[BugFix] Fix INSERT OVERWRITE BY NAME Fails with SQL Unknown Error

rohitrs1983 opened the pull request:
Fixes #58762 

InsertAnalyzer.analyzeWithDeferredLock() is invoked 2 times, on first invocation targetcolumns is null and this check is success; and targetcols is updated with cols specified in select list. When InsertAnalyzer.analyzeWithDeferredLock() is invoked second time for InertOverwriteJob, targetcolumns is not null, it contains the colums from select list. so this check will fail.

mysql> CREATE TABLE my_table (id VARCHAR(64) NOT NULL, uid BIGINT NOT NULL AUTO_INCREMENT) PRIMARY KEY (`id`);
Query OK, 0 rows affected (0.03 sec)

mysql> CREATE TABLE my_other_table (id VARCHAR(64) NOT NULL, uid BIGINT NOT NULL AUTO_INCREMENT) PRIMARY KEY (`id`);
Query OK, 0 rows affected (0.22 sec)

mysql> INSERT INTO my_other_table (id) VALUES ("o1");
Query OK, 1 row affected (0.58 sec)
{'label':'insert_6bb693d9-3897-11f0-8111-fe188ababd24', 'status':'VISIBLE', 'txnId':'23087'}

mysql> INSERT OVERWRITE my_table BY NAME SELECT id FROM my_other_table;
Query OK, 1 row affected (0.27 sec)
{'label':'insert_74dad179-3897-11f0-8111-fe188ababd24', 'status':'VISIBLE', 'txnId':'23089'}

mysql> select * from my_table;
+------+------+
| id   | uid  |
+------+------+
| o1   |    1 |
+------+------+
1 row in set (0.02 sec)


## What type of PR is this:

- [X] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [X] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [X] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [X] I have checked the version labels which the pr will be auto-backported to the target branch
  - [X] 3.5
  - [X] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1

alvin-celerdata commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/InsertAnalyzer.java:
```diff
@@ -242,7 +242,9 @@ public static void analyzeWithDeferredLock(InsertStmt insertStmt, ConnectContext
             }
 
-            Preconditions.checkState(insertStmt.getTargetColumnNames() == null);
+            if (insertStmt.isOverwrite()) {
+                Preconditions.checkState(insertStmt.getTargetColumnNames() == null);
+            }
```
Could you explain why this caused an Unknown Error before this change? 
It confuses me that the precondition of `insertStmt.getTargetColumnNames() == null` doesn't match.

rohitrs1983 replied:
@alvin-celerdata InsertAnalyzer.analyzeWithDeferredLock() is invoked 2 times, on first invocation targetcolumns is null and this check is success; when InsertAnalyzer.analyzeWithDeferredLock() is invoked second time for InertOverwriteJob, targetcolumns have the initialized the select columns. so this check will fail.

alvin-celerdata replied:
If this is the case, I would suggest removing this state check directly.
By the way, could you include this explanation in the commit message?

rohitrs1983 replied:
@alvin-celerdata removed the state check.

alvin-celerdata replied:
![image](https://github.com/user-attachments/assets/c8a6dc63-8f3d-41d9-95a3-6a2457c962ae)

Hi @rohitrs1983 what I mean is to add the explanation here in the photo, so that everyone can know it.

rohitrs1983 replied:
@alvin-celerdata updated.

rohitrs1983 committed:
InsertAnalyzer.analyzeWithDeferredLock() is invoked 2 times, on first invocation targetcolumns is null and this check is success; and targetcols is updated with cols specified in select list.
When InsertAnalyzer.analyzeWithDeferredLock() is invoked second time for InertOverwriteJob, targetcolumns is not null, it contains the colums from select list.
so this check will fail. there is no need of this state check, we can remove the state check entirely.

Signed-off-by: Rohit Satardekar <rohitrs1983@gmail.com>

alvin-celerdata commented:
@mergify rebase

alvin-celerdata commented:
@mergify rebase

rohitrs1983 committed:
InsertAnalyzer.analyzeWithDeferredLock() is invoked 2 times, on first invocation targetcolumns is null and this check is success; and targetcols is updated with cols specified in select list.
When InsertAnalyzer.analyzeWithDeferredLock() is invoked second time for InertOverwriteJob, targetcolumns is not null, it contains the colums from select list.
so this check will fail. there is no need of this state check, we can remove the state check entirely.

Signed-off-by: Rohit Satardekar <rohitrs1983@gmail.com>
